### PR TITLE
Inference & optimisation (+ visualisation) all in automated pipeline 🚀 

### LINKED ISSUE #100: [FEATURE REQUEST] Add options to specify graph NN type
Graph convolutional networks? Graph attention networks? Something else?

Could make this specifiable somewhere (probably in the config).

When implemented, please delete this: 
https://github.com/alan-turing-institute/grace/blob/7964502b1fe929ba3831305b3f85aea876c6f4ac/grace/training/README.md?plain=1#L113-L115

### LINKED ISSUE #232: Transform notebooks into proper pipeline in `run.py`
Add notebooks to the pipeline soon. All plots can be saved in the run directory (maybe under a subdirectory `plots`).  A function `evaluation(graph, classifier)` could be written more generally so it can be used to evaluate performance at epoch level (if that is useful), after training, as well as in an unseen graph.

_Originally posted by @crangelsmith in https://github.com/alan-turing-institute/grace/issues/229#issuecomment-1715705811_

### LINKED ISSUE #238: [DEVELOPMENT] Remove redundant `run_dir` argument from config
Remove only from the config.json when specifying the run, not from the `Config` object in `config.py` otherwise the run doesn't export the time-stamped directory.

### LINKED ISSUE #239: [DEVELOPMENT] Separate train & valid image datasets for GCN training
**Describe the issue here**

### LINKED ISSUE #240: [DEVELOPMENT] Change `transform` function to remove augmentations at valid / infer time
**Describe the issue here**

### LINKED ISSUE #241: [DEVELOPMENT] Define validation whole graph plotter param with certain frequency

### LINKED ISSUE #242: [DEVELOPMENT] Run inference on an unseen graph in `run.py`
Relates to #232 
Needs #241

Plot & save out the visualisations on validation graph with certain frequency

### LINKED ISSUE #243: [DEVELOPMENT] Calculate GNN perfomance metrics after the last epoch on valid image
- confusion matrices
- areas under curves
- final whole graph plot with highlighted node / edge probability score

### LINKED ISSUE #245: [DEVELOPMENT] Create a helper function to create datasets in `run.py`
Currently, the code gets replicated a lot. Take the transform & load_dataset functions outside of the `run.py` pipeline.

### LINKED ISSUE #246: [DEVELOPMENT] Serialise inference for more than just single graph

### LINKED ISSUE #248: [DEVELOPMENT] Add infer dataset performance to Tensorboard log
Re-considered, there's no need.

### LINKED ISSUE #249: [DEVELOPMENT] Store numerical performance metrics in `infer` subfolder
**Describe the issue here**

### LINKED ISSUE #251: [DEVELOPMENT] Animate the output of validation graphs

### LINKED ISSUE #252: [DEVELOPMENT] Optimise graphs & save out connected components fig
**Describe the issue here**

### LINKED ISSUE #253: [DEVELOPMENT] Update config.json + training subfolder README documentation

### LINKED ISSUE #256: [FEATURE REQUEST] Validate required config hyperparameters
Write a short sanity check to ensure all required directories were defined in the config.

### PR DESCRIPTION
**Pull request by KristinaUlicna**
## PR contribution summary

**Why is this PR useful / good for?** *Please describe the problem(s) you're trying to address.*
* Inference visualisation inside the `run.py` to automate the entire pipeline
* [WIP]

### List of proposed changes / linked issues & discussions

- [x] Resolves #100 
- [x] Resolves #232
- [x] Resolves #238 
- [x] Resolves #239 
- [x] Resolves #240 
- [x] Resolves #241 
- [x] Resolves #242 
- [x] Resolves #243 
- [x] Resolves #245  
- [x] Resolves #246 
- [x] Resolves #251 
- [x] Partially resolves #249 
- [x] Resolves #252
- [x] Resolves #253
- [x] Resolves #256
- [ ] 🏷️ Addresses (but does not close) #247 

### What should a reviewer concentrate their feedback on?

+ 🏃 Scripts / notebooks to run
+ 💻 Code quality
+ ✍🏻 Spell check
+ 📝 Everything looks OK?

#### What type of PR is this? (check all applicable)

+ 🪄 Feature
+ 🐛 Bug fix
+ 📝 Documentation / update README.md
+ #️⃣ Documentation / code annotation
+ 🧑‍💻 Code refactor / style
+ 🔥 Performance Improvements
+ ✅ Test

#### Added tests?

+ 👍 yes
+ 🙅 no, because they aren't needed
+ 🙋 no, because I need some help

---

## PR review summary

Describe what this PR does & **how you reviewed** the individual items, where needed:

* <TEXT>

*Some helper checks to tick off:*

* Focus on image annotation
* Focus on model training
* Could any optimization be applied?
* Is there any redundant code?
* Are there any spelling errors?

**In conclusion**, after my review, I'd like to:

* 🙋 ask some clarifying questions
* 🙅 suggest some specific changes

### CODE REVIEW

**Comment by KristinaUlicna** [comment]
Tests pass locally but keep failing on GitHub as the environment running the tests doesn't have `magicgui` module imported. This is a temporary hack while I'm getting the inference to run properly. No panic 😄 🚧

**Comment by KristinaUlicna** [comment]
Edits after branch rebase & code refactoring to allow smooth running of the entire pipeline 🚀

File: grace/run.py [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+    # Save the trained model:
+    model_save_fn = run_dir / "classifier.pt"
+    torch.save(classifier, model_save_fn)
+    write_config_file(config, filetype="json")
+    write_config_file(config, filetype="yaml")
+
+    # Project the TSNE manifold:
+    if config.visualise_tsne_manifold is True:
+        # TODO: Implement
+        pass
```
Comment (crangelsmith) [inline grace/run.py:173]: maybe add a NotImplemented warning here?

File: grace/run.py [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+
+    # Animate the validation outputs:
+    if config.animate_valid_progress is True:
+        animate_entire_valid_set(run_dir / "valid", verbose=False)
+
+    # Run inference on the final, trained model on unseen data:
+    GLP = GraphLabelPredictor(model_save_fn)
```
Comment (crangelsmith) [inline grace/run.py:180]: Thinking about what @quantumjot mentioned today about separating training from inference, this is exactly where this could happen. 

To maintain the flexibility of the run script in the short-term we could say if a `model_save_fn` is given in the config file (and maybe an inference-only flag), then run script from here. Further in the line (when we know Grace is working as expected) we can think how es best to abstract each piece.

This comment is not for this PR, just to keep this in mind for the next step.

File: grace/training/train.py [hunk 3]
```diff
@@ -215,5 +222,32 @@ def valid(loader):
         # Print out the logging string:
         LOGGER.info(logger_string)
```
Comment (crangelsmith) [inline grace/training/train.py:223]: could we add a line break at the end of each epoch string? this will make the outputs easier to read.

File: grace/training/config.py [hunk 4]
```diff
@@ -96,52 +110,113 @@ def load_config_params(params_file: Union[str, Path]) -> Config:
...
+) -> None:
+    if isinstance(filename, str):
+        filename = Path(filename)
+
+    if filename.suffix == ".json":
+        with open(filename, "w") as outfile:
+            json.dump(parameters_dict, outfile, indent=4)
+
+    elif filename.suffix == ".yaml":
+        with open(filename, "w") as outfile:
```
Comment (crangelsmith) [inline grace/training/config.py:203]: If I try to run the pipeline starting for the output yaml for a previous run, the code breaks because I think the outputs are changed (e.g lists turned to strings).

This is not the case with the output json file which is consistent.

**Comment by crangelsmith** [comment]
I've run the code successfully with both json and yaml file, except for a couple of minor comments about the code looks good to me!

File: grace/run.py (hunk 1 above)
Comment (KristinaUlicna) [reply grace/run.py:173]: Coming in the next commit 🚀

File: grace/training/train.py (hunk 3 above)
Comment (KristinaUlicna) [reply grace/training/train.py:223]: Raising a new issue from this to further customise the logger in the `styling.py` script.

File: grace/training/config.py (hunk 4 above)
Comment (KristinaUlicna) [reply grace/training/config.py:215]: This should now be fixed in the new commit wave! Thank you for the thorough checks - I didn't spot that the params do not parse correctly 🚀

**Review by crangelsmith — APPROVED** [review]
